src/main/java/org/beanpod/switchboard/controller/DeviceController.java

```diff
@@ -10,65 +11,81 @@
-    }
-    return ResponseEntity.ok(device.get());
+  @Override
+  public ResponseEntity<DeviceModel> retrieveDevice(@PathVariable String serialNumber) {
+    DeviceDto deviceDto =
+        service
+            .findDevice(serialNumber)
+            .orElseThrow(() -> new ExceptionType.DeviceNotFoundException(serialNumber));
+
+    return Optional.of(deviceDto)
+        .map(deviceMapper::toDeviceModel)
+        .map(ResponseEntity::ok)
+        .orElseThrow(() -> new ExceptionType.UnknownException(CONTROLLER_NAME));
   }
 
-  @PostMapping
-  public ResponseEntity<DeviceDto> createDevice(@RequestBody @Valid DeviceDto device) {
-    Optional<DeviceDto> deviceLookup = service.findDevice(device.getSerialNumber());
+  @Override
+  public ResponseEntity<DeviceModel> createDevice(@Valid CreateDeviceRequest createDeviceRequest) {
+    Optional<DeviceDto> deviceLookup = service.findDevice(createDeviceRequest.getSerialNumber());
     if (deviceLookup.isPresent()) {
-      throw new ExceptionType.DeviceAlreadyExistsException(device.getSerialNumber());
+      throw new ExceptionType.DeviceAlreadyExistsException(createDeviceRequest.getSerialNumber());
     }
-    return ResponseEntity.ok(service.save(device));
+
+    return Optional.of(createDeviceRequest)
+        .map(createRequest -> service.createDevice(createRequest, request.getRemoteAddr()))
+        .map(deviceMapper::toDeviceModel)
+        .map(ResponseEntity::ok)
+        .orElseThrow(() -> new ExceptionType.UnknownException(CONTROLLER_NAME));
   }
 
-  @DeleteMapping("/{serialNumber}")
+  @Override
   @Transactional
   public ResponseEntity<String> deleteDevice(@PathVariable String serialNumber) {
     Long response = service.deleteDevice(serialNumber);
     if (response != 1) {
       throw new ExceptionType.DeviceNotFoundException(serialNumber);
     }
-    return ResponseEntity.ok("Device with serial number " + serialNumber + " Deleted");
+    return ResponseEntity.ok("Device with serial number " + serialNumber + " deleted");
   }
 
-  @PutMapping
+  @Override
   @Transactional
-  public ResponseEntity<DeviceDto> updateDevice(@RequestBody @Valid DeviceDto device) {
-    Optional<DeviceDto> deviceLookup = service.findDevice(device.getSerialNumber());
-    if (deviceLookup.isEmpty()) {
-      throw new ExceptionType.DeviceNotFoundException(device.getSerialNumber());
-    }
-    return ResponseEntity.ok(service.save(device));
+  public ResponseEntity<DeviceModel> updateDevice(@Valid DeviceModel deviceModel) {
+    DeviceDto deviceDto =
+        service
+            .findDevice(deviceModel.getSerialNumber())
+            .orElseThrow(
+                () -> new ExceptionType.DeviceNotFoundException(deviceModel.getSerialNumber()));
+
+    return Optional.of(deviceDto)
+        .map(service::save)
+        .map(deviceMapper::toDeviceModel)
+        .map(ResponseEntity::ok)
+        .orElseThrow(() -> new ExceptionType.UnknownException(CONTROLLER_NAME));
```

Comment:
Regression: This code for the update device endpoint is broken, you simply re-save what was retrieved in the database, and don't use the deviceModel parameter. So nothing is done. I think you meant to map the incoming deviceModel to a DTO then save it.